### DISCUSSION TITLE
Factor of 1.5 between errors from iPidControl::getErrors and actual errors

### DISCUSSION THREAD

**Discussion by hu-yue** [robotology/community#154] [Q&A] [answered]
Hi,

I recorded the data from `iPidControl::getErrors` and then also data from `iEncoders`, then I computed the errors by `qdes-iEncoders`. I have realised that the errors from `iPidControl::getErrors` are of a factor of circa `1.5` smaller than the ones I computed. By dividing `qdes-iEncoders` by `1.5` they are basically the same as the ones from `iPidControl::getErrors`.
See plots below:
In blue errors from `iPidControl::getErrors` and in red `qdes-iEncoders`:
![error1](https://cloud.githubusercontent.com/assets/7060968/19812417/afad7e0a-9d35-11e6-8f38-69af27315d33.png)

In blue errors from `iPidControl::getErrors` and in red `(qdes-iEncoders)/1.5`:
![error2](https://cloud.githubusercontent.com/assets/7060968/19812420/b4fdce50-9d35-11e6-9411-6822e3b69004.png)

I thought that the errors from `iPidControl::getErrors` are the tracking errors, as this is also what the `controlBoardDumper` uses to return the position errors.

The questions are:
- Am I doing something wrong?
- Is this factor really there?
- If yes, what is it and why?

Thanks!

**Comment by traversaro** [comment]
Hi @hu-yue , 
   on which joints did you get this data?

**Comment by hu-yue** [comment]
@traversaro these specific plots are for the left ankle roll, however I get the same also on left knee and right knee, I can also test other joints.

**Comment by hu-yue** [comment]
It seems to be true for all joints (of iCubHeidelberg, so two legs and torso), however on the torso the tracking is quite bad, so I'm not sure about the outcome.

**Comment by hu-yue** [comment]
I tried to understand if this was due to the latest firmware and software update, since I had data recorded from before. I have to say that I couldn't actually understand it. The computed errors and the errors from `iPidControl::getErrors` were different, but by dividing the computed errors by 1.5 they did not match with `iPidControl::getErrors`.

In general, I'd like to go back to my initial questions:
- Why are these errors different? Aren't they supposed to be the same or I'm not understanding what are the values from `iPidControl::getErrors`?

**Comment by randaz81** [comment]
afaik, they should be identical on all joints of your robot. Suspected bug.

**Comment by hu-yue** [comment]
@randaz81 @ale-git ok, I wait for further updates then

@KevinAStein FYI

**Comment by ale-git** [comment]
Hi @hu-yue , which controlmode has been used for the test? In the plots, is the x axis the time in milliseconds? The errors are measured in degrees? Thanks.

**Comment by hu-yue** [comment]
@ale-git 
- the control mode is `position direct`
- the x axis is the number of samples, the total time is 6 sec, and I was sending commands with a thread rate of 10 ms
- the errors are in degrees

**Comment by hu-yue** [comment]
Ok sorry the statement about the time is clearly wrong, total time is 6 sec, so the x axis is just number of samples

**Comment by ale-git** [comment]
Hi @hu-yue , you should consider that there is a transmission delay Tt between your qdes at high level and the qdes_pid inside the pid controller, so qdes - iEncoders is in general different from iPidControl::getErrors = qdes_pid - iEncoders. If the matching between the qdes_pid reference and the joint position is affected by a Tc delay related to the pid time constant, then you have:
qdes = A \* sin((2_pi/6) \* t)
qdes_pid = A \* sin((2_pi/6) \* (t-Tt))
qenc = A \* sin((2_pi/6) \* (t-Tt-Tc))
Thus
qdes - qenc  =(about) A \* (2_pi/6) \* (Tt + Tc) \* cos((2_pi/6) \* t)  (as shown by the plot)
qdes_pid - qenc =(about) A \* (2_pi/6) \* Tc \* cos((2*pi/6) \* t)  (as shown by the plot)
and finally 
(qdes - qenc) / (qdes_pid - qenc) = (Tt + Tc) / Tc 
Given your software configuration, can the transmission delay explain the gain factor?

**Comment by ale-git** [comment]
If the above hypothesis is correct, you should read a difference between qdes and iPidControl::getReference.

**Comment by pattacini** [comment]
Hi @hu-yue @ale-git 

The acquisitions seem to refer to a steady-state condition where the error is quite low (~`1-2 deg`).
I would suggest to repeat the experiment logging all the involved quantities during a reasonable transition of `qdes` (e.g. from `0 deg` to `40-50 deg`).

This should allow us to better test the hypothesis and appreciate possible latency among these variables.

**Comment by ale-git** [comment]
Just an example with Tc = 0.01 and Tt = 0.005:

![poserrors](https://cloud.githubusercontent.com/assets/4729270/20001715/bbec3e66-a27d-11e6-8ba6-f7bacc3f576b.png)

**Comment by hu-yue** [comment]
@ale-git @pattacini thanks for the replies!

I run another experiment on the left knee joint performing a cos wave that moved the joint from `0` to `-90` and back in `2 [sec]`, the thread was running within the specified rate, so no delay should have been introduced by the thread itself. 
I think the transmission delay can explain the difference between the two errors.
The data from `iPidControl::getReferences` seem to be delayed respect to `qDes` and then the encoders have a higher delay time.

Here are plots of the case I tested:

Reference trajectory (qDes), encoders (Enc) and `iPidControl::getReferences` (Pos ref)
![reference](https://cloud.githubusercontent.com/assets/7060968/20013390/3b44b622-a2b3-11e6-9297-119d7d54f44d.png)

Errors:
![errors](https://cloud.githubusercontent.com/assets/7060968/20013397/42d8acea-a2b3-11e6-8800-842f8fb38f8b.png)

Then I suppose that the actual execution errors should be the ones returned by `iPidControl::getErrors` and one should not compute `qDes - iEncoders`, as this would increase the errors due to the delay factor?

**Comment by ale-git** [comment, marked as answer]
Hi @hu-yue, yes, the error you should look at is returned by iPidControl::getErrors because it is synchronized to the encoder data. The use of qDes - iEncoders introduces a gain factor due to the delay between the qDes and the low level joint position reference.
Cheers,
Alessandro

**Comment by hu-yue** [comment]
@ale-git ok perfect, thanks for all the explanations! Now things are clearer.

**Comment by pattacini** [comment]
Thanks @ale-git for the great handling of this issue 👍 

@hu-yue I'd just like to add a brief comment upon closure from a bit higher standpoint.
Taking `qdes-enc` as a measure of the error remains perfectly viable though, given that controllers can be easily designed to account for latency too, of course depending on the requirements the designer has in mind.

**Comment by randaz81** [comment]
iPidControl::getErrors() has been just implemented in gazebo, see https://github.com/robotology/gazebo-yarp-plugins/issues/260
